### DISCUSSION TITLE
MTB and Suunto : fit export or workout upload

### DISCUSSION THREAD

**Discussion by mff73** [mytourbook/mytourbook#1194] [Q&A] [answered]
Hello,
I am struggling with the following issue since a long time and i can't figure where the issue is : in MTB or in Suunto, or both.
Here are two files and two screen recordings for the same activity, originally recorded by Suunto Device.
One file is the original activity directly from Suunto (orginal fit file), the other is the exact same activity, but imported to MTB and just re-uploaded to SuuntoApp without any change.

This activity contains a pause, and SuuntoApp is not "displaying" properly the MTB reuploaded one : either in the activity movie replay, and in the detailed graphs.
The global GPS track is correct, but during and after the pause, there seems to be some mixing of timestamps or whatever i am not able to understand myself.

https://github.com/mytourbook/mytourbook/assets/50132146/72b5870b-34fc-4d5f-83b4-11b4b297ad6a

[2023-09-23T14-51-Original Suunto.zip](https://github.com/mytourbook/mytourbook/files/12839962/2023-09-23T14-51-Original.Suunto.zip)

https://github.com/mytourbook/mytourbook/assets/50132146/0bcc087b-c779-40c1-86e7-4f942c2b8f1a

[2023-09-23T12-51-export MTB.zip](https://github.com/mytourbook/mytourbook/files/12839965/2023-09-23T12-51-export.MTB.zip)

The track movie replay is not playing until the end of the track, but is stopping before the end, and from my analysis, it stops exactley at the "pause duration" before the end.

I decoded both fit files, and tried a lot of manual fixes, but i was not able to "clean" the MTB one to be properly managed by SuuntoApp.

I must precise that, re-uploading the orginal suunto fit file into SuuntoApp, gives a correct Suunto activity. (i am using runalyze to do this directly from fit file to Suunto). Which might mean that something in MTB fit file is wrong (no ?).

Issue in MTB fit file ? Issue in Suunto ? Both ? I search alone and couldn't find the answer.
Thanks for any help, and if one need more clarification or data, don't hesistate.

**Comment by FJBDev** [comment]
So what you are observing is that if you delete that pause, then the movie replay is playing until the end of the track ?

>The global GPS track is correct, but during and after the pause, there seems to be some mixing of timestamps or whatever i am not able to understand myself.

Did you look and found some erroneous timestamps ?

EDIT: And if an activity has several pauses, does the movie player stop at the first pause ? the last pause ?

**Reply by mff73** [reply]
If i remove the pause, yes, the movie replay until the end of the track, but, then the total recorded time is wrong (which is normal, but wrong). Maybe also link to the fact that MTB seems to create "fake" GPS points during pause (like a straight line between beginning of pause and end of pause).

There is no erroneous timestamps in the MTB fit file : same timestamps for same GPS points as in Original file, but a "wrong" replay in Suunto movie.

**Reply by FJBDev** [reply]
> Maybe also link to the fact that MTB seems to create "fake" GPS points during pause (like a straight line between beginning of pause and end of pause).


MT doesn't create fake GPS point. The straight line you see if only a display addition that doesn't get added in the original tour.

**Reply by mff73** [reply]
maybe my wording with fake is not correct, but in my examples, 
MTB is having GPS lat&lon for all timestamps during the pause. 
![image](https://github.com/mytourbook/mytourbook/assets/50132146/c4797899-e584-4209-bc84-8fcc3c6fd92c)

and the fit file export also
Here timestamp of beginning of the pause
![image](https://github.com/mytourbook/mytourbook/assets/50132146/2deee6c1-f6ab-4008-9924-c166a73a6d44)

And records after this pause start.
![image](https://github.com/mytourbook/mytourbook/assets/50132146/c0a58c5c-b842-4304-9a83-18c437ebe360)

It is these GPS values which i am calling 'fake', as, according to the fit export, it is "like" there is a track during the pause.
Maybe it is this workaround which enable to display a straight line.



While in the suunto fit file, there is not lat/lon during the pause.
![image](https://github.com/mytourbook/mytourbook/assets/50132146/d836ee8d-6a60-4206-bb76-a50ce2de39aa)



I am not sure it is the issue, because, during my tests, i simulated an MTB fit file, with removing lat/lon during the pause --> issue is still the same in SuuntoApp.
I am quite really lost :)

**Reply by FJBDev** [reply]
> > and the fit file export also Here timestamp of beginning of the pause

Then maybe it's me who's wrong, it wouldn't be the first of the last time 😄 

I need to dig deeper into your files. I see that MT generates the pause but not sure why Suunto doesn't like it

**Comment by FJBDev** [comment]
I tried to see if putting the pauses during the activity in the FIT file instead of at the end would change something but no luck. I need to investigate more....

**Reply by mff73** [reply]
Thanks and good luck, I searched a lot, and didn't understand why suunto's fit file is well accepted back into Suunto App, and not MTB 's one.
But maybe something specific in Suunto itself.
I tried to ask in their forum, but it is so a specific case....
Question : don't you have any "support" contact because you certainly asked to Suunto the authorization for MTB to upload files?

**Reply by FJBDev** [reply]
>I tried to ask in their forum, but it is so a specific case....

Did you get any answers/hints ?

>don't you have any "support" contact because you certainly asked to Suunto the authorization for MTB to upload files?

No, they don't offer any support. And my experience with Suunto support is pretty poor.

**Reply by mff73** [reply]
OK, I will email one of their backend developer, to submit him MTB fit file. I personnaly tried almost anything which seems logic.
Both files are really quite similar.

**Reply by FJBDev** [reply]
>OK, I will email one of their backend developer, to submit him MTB fit file.

Great, let me know what they say. I will still continue to do some testing but I haven't found anything yet. It could be an issue on how MT is generating the FIT file but could it also be an issue in the Suunto App ?

**Reply by FJBDev** [reply]
Also, for what it's worth, I've tried to upload the same activity in FIT generated by GoldenCheetah and Suunto doesn't even accept it

` Error while uploading the workout of Id "652838bae48d0556f9d084e4": "Internal error"`

**Reply by mff73** [reply]
for golden cheetah's fit file, eventually try to process them first with the ActivityRepairTool.jar included in fit SDK csvtools.

**Reply by FJBDev** [reply]
> for golden cheetah's fit file, eventually try to process them first with the ActivityRepairTool.jar included in fit SDK csvtools.

I think in the end it's not useful because GoldenCheetah doesn't seem to support pauses in FIT files

![image](https://github.com/mytourbook/mytourbook/assets/25235665/8e1ea051-05bf-4494-9763-efd49ea53da3)

**Reply by FJBDev** [reply]
Another thing you could do to help troubleshoot it. Can you create a very small tour that reproduces the issue so that it's easier to identify the potential differences between the original one and the MT one ? (aka: less lines in the CSV makes it easier to analyze it).

**Reply by mff73** [reply]
Hi, sorry, little busy these days.
Here is this second test activity, a short one, done in the middle of nowhere for this test :)  
![image](https://github.com/mytourbook/mytourbook/assets/50132146/43258ab7-3b08-4b7f-8445-f368a7017d27)

1min walk in one direction, 33sec of pause, and walk back to start.

Here are the suunto original file, and the mtb exported one.
Issue is the same in suunto replay : stopped somewhere 33s before the end of the GPS track

[2023-10-19T16-15-suunto & MTB fit file issue with pause.zip](https://github.com/mytourbook/mytourbook/files/13045169/2023-10-19T16-15-suunto.MTB.fit.file.issue.with.pause.zip)

[EDIT]
why GPS coordinates are not the same exactely from original file (on the left) and in MTB fit file (on the right) for the exact same timestamps (here are just the first 3 seconds ?
![image](https://github.com/mytourbook/mytourbook/assets/50132146/599105e5-d60a-4029-a78d-a5593431ac62)

**Reply by FJBDev** [reply]
> 1min walk in one direction, 33sec of pause, and walk back to start.
> 
> Here are the suunto original file, and the mtb exported one. Issue is the same in suunto replay : stopped somewhere 33s before the end of the GPS track
> 

Superb, thanks, I will investigate that soon.
First thing I notice when playing it with the Suunto App is that the replay stops after the pause and not at the pause. Do you notice the same thing ? (maybe the issue is not related to pauses ?)

> 
> [EDIT] why GPS coordinates are not the same exactely from original file (on the left) and in MTB fit file (on the right) for the exact same timestamps (here are just the first 3 seconds ? ![image](https://user-images.githubusercontent.com/50132146/276691947-599105e5-d60a-4029-a78d-a5593431ac62.png)

Hmmm, interesting, I will investigate that as well. Maybe related to how much precision is saved in the db ? (just a hunch though...)

**Comment by FJBDev** [comment]
Also, ive tried both of the activities in Relive and the replay works fine even with the activities generated by MT. Attached is an example with your latest activity.

https://github.com/mytourbook/mytourbook/assets/25235665/31a890b3-297f-427b-9c41-81bd6208bb10

**Comment by FJBDev** [comment]
>OK, I will email one of their backend developer, to submit him MTB fit file

Did you get any response from the Suunto developer ? I'm never holding my breath with Suunto but you never know !

**Comment by FJBDev** [comment]
After doing more tests, it seems to me that the replay stops because the player might use the recorded time vs elapsed time.
If you put a pause at the beginning, the replay stops at the end.

**Comment by mff73** [comment]
>First thing I notice when playing it with the Suunto App is that the replay stops after the pause and not at the pause. Do you notice the same thing ? (maybe the issue is not related to pauses ?)


> After doing more tests, it seems to me that the replay stops because the player might use the recorded time vs elapsed time. If you put a pause at the beginning, the replay stops at the end.

Replay seems stucked exactely as if all track points (GPS) are replayed with timestamps of recorded time and not elapsed time, yes, i think the same.
Could be linked to the fact that MTB fit file is containing GPS position during the pause elapsed time, PLUS, Suunto not managing this correctly. Because i remember trying to remove these GPS locations in MTB file manually, but no effect.
Suunto seems to be able to handle its own fit file to be reimported, but not MTB's one :(


> Did you get any response from the Suunto developer ? I'm never holding my breath with Suunto but you never know !
Not yet

Btw, thanks for your time spent on this :)

**Comment by FJBDev** [comment]
>Because i remember trying to remove these GPS locations in MTB file manually, but no effect.

Correct. I did some testing myself and modified the code to remove those points but it didn't change anything for the replay.
I think having someone at Suunto look at it will definitely help us

>Btw, thanks for your time spent on this :)

You're welcome!

**Reply by mff73** [reply]
Hi Fred,
I still don't have any information from Suunto dev, but i still checking things, and didn't find any way to make it work.
But i see something strange.
In fact i am decoding fit files with FITtoCSV using FitCSVtool.jar from the Fit SDK (sorry, i am just a geek user not any coder :) ).
With default command i always used, i didn't see the following, but using -s option, meaning decoding also invalid fields, i see this : 

![image](https://github.com/mytourbook/mytourbook/assets/50132146/674dcab6-213a-4e34-9732-8410f9fbecc6)

on the left, an MTB fit export without -s option and on the right with -s option : 
timer_trigger data seems wrong in fact if displayed correctly (green box), while they should only be 0 or 1
why i didn't see this before : there is nothing in the activity section : empty, just the message without any values.

i fixed that manually, but it didn't change anything, but it could be worth checking this if you want.

Last point, it seems to be missing a few comas ( , ) in the Definition of records, to properly define the number of fileds including the extra : enhanced....values
![image](https://github.com/mytourbook/mytourbook/assets/50132146/3116957e-c75c-4959-a3fd-57375d805c18)

Normal ? or a glich from decoding ?

**Comment by FJBDev** [comment]
> Hi Fred, I still don't have any information from Suunto dev, but i still checking things, and didn't find any way to make it work. But i see something strange. In fact i am decoding fit files with FITtoCSV using FitCSVtool.jar from the Fit SDK (sorry, i am just a geek user not any coder :) ). With default command i always used, i didn't see the following, but using -s option, meaning decoding also invalid fields, i see this :

Can you ask Suunto again ? But maybe it's not an important issue for them I guess ?

> 
> 
> on the left, an MTB fit export without -s option and on the right with -s option : timer_trigger data seems wrong in fact if displayed correctly (green box), while they should only be 0 or 1 why i didn't see this before : there is nothing in the activity section : empty, just the message without any values.
> 
> i fixed that manually, but it didn't change anything, but it could be worth checking this if you want.

Thanks for your investigation. I looked in the code and I couldn't find anywhere where `time_trigger` is set. We set the start time but i think it's a different datapoint.

> 
> Last point, it seems to be missing a few comas ( , ) in the Definition of records, to properly define the number of fileds including the extra : enhanced....values 
> 
> Normal ? or a glich from decoding ?

I'm not sure about the commas since they're written when decoding from the FIT to CSV file and I am not sure what in the SDK would output them vs not. Also, does that change something if you add the commas ?

**Comment by mff73** [comment]
> Can you ask Suunto again ? But maybe it's not an important issue for them I guess ?

I am doing it almost monthly ;) It is probably not an important issue, yes. It also may not be an important issue for MTB, i really don't know in fact, without knowing the root cause of it. 3D replay is not a so important feature in Suunto's world to me, but, it may be some revelative of the fact that potentially, MTB fit files are not super perfect :) and may cause any issues in any future.

> Thanks for your investigation. I looked in the code and I couldn't find anywhere where `time_trigger` is set. We set the start time but i think it's a different datapoint.

You surely know it 😁, but timer_trigger is the pause status : manual/automatic.
Maybe decoding files with fitcsvtool is just not showing the reality of the fit content.
But maybe the procedures that MTB is using to generate fit files are also containing some "flaws" which are generating issues in the resulting file, which suunto is not able to manage.
If MTB is not adding timer_trigger itself, maybe it is added (at encoding stage, or just for display, at decoding stage), with wrong values.
What about adding them, officially, for any pause (as MTB is knowing if a pause is manual or automatic, no ? )
Look at fitfileviewer "decoding" comparison between MTB and Suunto file
![image](https://github.com/mytourbook/mytourbook/assets/50132146/ae051f13-0508-4b9e-8d41-a3b2b67a4ecc)


> I'm not sure about the commas since they're written when decoding from the FIT to CSV file and I am not sure what in the SDK would output them vs not. Also, does that change something if you add the commas ?

No :) it is also my new understanding : what i am seeing in decoding a fit file, is just a csv visualization of the fit content, so probably not the real reality truth. But i am thinking naively, that is official SDH csv decoder is showing these, except some bugs, there should be a strange reality in the files. Maybe i could ask in sdk forum, just to learn more about this.


Very last point, i installed Fitfilerepair tool demo version, and importing an MTB fit file gives : 
![image](https://github.com/mytourbook/mytourbook/assets/50132146/5c7ab7c7-a7bb-4f61-8463-0bdf97aba9db)

and

![image](https://github.com/mytourbook/mytourbook/assets/50132146/d25f4c9c-34c6-4ae0-a668-621d00c780b3)

another clue of something weird, isn't it ?

**Comment by FJBDev** [comment]
> I am doing it almost monthly ;)

Good. Let me know when they have answered, I wonder what their analysis will reveal

> it may be some revelative of the fact that potentially, MTB fit files are not super perfect :) and may cause any issues in any future.

Ok but what is a "perfect" fit file ? 
When I implemented, the MT FIT export, I followed the [official Garmin documentation](https://developer.garmin.com/fit/cookbook/encoding-activity-files/), I would like to think that it is the right documentation to follow, no ?
The MT Fit Export is used for the Strava Upload, Suunto Upload and while I am willing to fix any possible and arising issues, I am not aware of any other issues than this one for which we don't have any clues (yet) if it's  a problem in MT or in Suunto (afterall, my test with Relive worked successfully. I am curious if the issue appears when using the [Strava Flyover ](https://www.dcrainmaker.com/2023/11/strava-launches-activity-flyover.html)but I can't test myself because I don't have a Premium subscription)
Also, since you mention "perfect" fit file, in my experience, the Suunto files are far from perfect too and I think you discovered that recently [here](https://github.com/mytourbook/mytourbook/discussions/1284#discussioncomment-8419407) right ? Since I have used the FIT files from Suunto, they have never bothered to provide a file serial number, why ?
`[FIT] /home/frederic/Downloads/2-10-24 9:56 AM_09:56.fit - File serial number is missing, device id cannot not be set`

> 
> You surely know it 😁, but timer_trigger is the pause status : manual/automatic. 

No, `time_trigger` is not the pause status, `event_type`(0 vs 1) is, see [here](https://github.com/mytourbook/mytourbook/blob/7ff0e4b1cafb942e43ddf0d09ce092442f9900b2/bundles/net.tourbook.tests/src/test/export/files/Bye_bye_Silverton.csv#L561) for an example

> What about adding them, officially, for any pause (as MTB is knowing if a pause is manual or automatic, no ? ) 

I already added the[ pause type](https://github.com/mytourbook/mytourbook/pull/1230) in the FIT file export in 24.1

```
Fit Export (Frederic)

    Added the pause types

```

>  Maybe i could ask in sdk forum, just to learn more about this.

Good idea. Feel free to ask any questions you have about the SDK or the FIT file in the [official forum](https://forums.garmin.com/developer/fit-sdk/f/discussion). It seems to me that they are pretty responsive (at least better than Suunto...)
> 
> Very last point, i installed Fitfilerepair tool demo version, and importing an MTB fit file gives : 
> 
> another clue of something weird, isn't it ?

I think that's normal because MyTourbook is not a device nor a product that is officially registered (unlike a public watch or GPS unit for example) and hence we only output this information.
See [here](https://github.com/mytourbook/mytourbook/blob/7ff0e4b1cafb942e43ddf0d09ce092442f9900b2/bundles/net.tourbook.export/src/net/tourbook/export/fit/FitExporter.java#L88) when creating a FIT file

**Reply by mff73** [reply]
Sorry if i may have created any offense with my wordings :( it seems i am not using all the time the exact terminology to express my non expert feelings.

"Pause status" for me, was what you are calling "Pause type"
And it is effectively not totally the timer_trigger.
One PauseType is auto or manual, depending on two timer_trigger events, which are either manual or auto trigger.
![image](https://github.com/mytourbook/mytourbook/assets/50132146/be53622d-7601-4eed-8a58-9cf3e85df45d)

And comparing MTB 24.1 and Suunto's file, with fitfileviewer.com, it shows a difference, so there seems to be a difference somewhere

![image](https://github.com/mytourbook/mytourbook/assets/50132146/66d5a63b-a890-48e6-967d-bec068d585ea)

MTB seems to set timer_trigger = 0 (manual) for the manual stop, but there seems to be no event_data for the start.

I will ask to sdk forum what they are thinking and push to suunto to try to explain (they just may have an issue on their side).
Thanks for you time and patience

**Comment by mff73** [comment]
Hi,
I may have found what is not well accepted at suunto level for the 3d replay.
The timer start/stops events are placed at the very end of the fit file (at least in the csv decoded file).
And if putting each start/stop at its chronological place in the file, all seems accepted and working.
Current 24.1 export : at the end of the file, just before session and activity messages
![image](https://github.com/mytourbook/mytourbook/assets/50132146/8568adf9-c4b2-421c-87a1-04f25abf3aa4)

if inserting these events at their right timestamp place : left original file, right modified
![image](https://github.com/mytourbook/mytourbook/assets/50132146/b74d2801-51ad-4105-95e9-8d135dc1ab4b)


[pauses test fit files.zip](https://github.com/mytourbook/mytourbook/files/14318441/pauses.test.fit.files.zip)

Here are two fit files, one from MTB direct export and the one just tweaked to make it work with suunto.

**Comment by mff73** [comment]
Hello,
I am quite happy :) i can now manually "EDIT" MTB file files to make them work with Suunto.
The solution in my previous message seems to work for all my tests.
To simplify my manual job :) , instead of definining event messages each time and back to define record messages, i make the event message definition with another Definition number , exemple : 
First place = initial Start 
![image](https://github.com/mytourbook/mytourbook/assets/50132146/7ef794b6-74ee-4460-aa7a-c2c6d2b77b43)

Second place : first stop
![image](https://github.com/mytourbook/mytourbook/assets/50132146/febfcb53-ca16-4096-a73c-d793cb98bb42)

and so on 👍 

[EDIT]
Fit SDK forum answer:
![image](https://github.com/mytourbook/mytourbook/assets/50132146/68494ac0-d55d-4058-a0fd-eb673c72b695)

**Comment by FJBDev** [comment]
> Hi, I may have found what is not well accepted at suunto level for the 3d replay. The timer start/stops events are placed at the very end of the fit file (at least in the csv decoded file). And if putting each start/stop at its chronological place in the file, all seems accepted and working. Current 24.1 export : at the end of the file, just before session and activity messages ![
> 
> [pauses test fit files.zip](https://github.com/mytourbook/mytourbook/files/14318441/pauses.test.fit.files.zip)
> 
> Here are two fit files, one from MTB direct export and the one just tweaked to make it work with suunto.

Nice find ! Did Suunto answer your question to help you identify the issue ?

Once I am done with [this feature](https://github.com/mytourbook/mytourbook/issues/1280), I will work on making a fix to MT so that you don't have to fix each file manually.

**Comment by mff73** [comment]
> Nice find ! Did Suunto answer your question to help you identify the issue ?

Thanks, but no help yet from Suunto 😒, but i send them my findings, at least for them to validate it is well something known.
But, FitSDK answer seems to validate on fit file side.
Let see.

> Once I am done with [this feature](https://github.com/mytourbook/mytourbook/issues/1280), I will work on making a fix to MT so that you don't have to fix each file manually.

thanks 👍 a lot. No emergency, do not stop anything ongoing.
And if you don't care, i am also preparing some nice other findings about MTB's fit files, to make them little more compliant with FIT SDK (i don't say perfect 🥳😊). 

Again, to all MTB developpers, big Thanks to get this software so open to changes and tweaks 🙏

**Comment by mff73** [comment]
Hi,
To conclude my analysis about fit files export from MTB, i discover a new "tool" in the fit SDK : FitTestTool.jar
I submitted MTB fit file to it, and it founds 4 main "problems" .
Those problems seems currently not making any issues with Suunto Import for my case, but it may be worth having a look at them, to decide if it worth MTB to try to align.
Here are those "problems" : 
![image](https://github.com/mytourbook/mytourbook/assets/50132146/9455db68-8680-40f5-966a-29920b23aaf0)


1. FileId message : no manufacturer id 
![image](https://github.com/mytourbook/mytourbook/assets/50132146/48a29788-5e1f-46ff-94e4-ebb795ae41a9)

2. Activity message : no timestamp, no local timestamp 
3. Activity message : no total_timer_time 
![image](https://github.com/mytourbook/mytourbook/assets/50132146/a30d539c-6fc6-447d-a234-92be18ae4d37)

4. Session message : no timestamp
![image](https://github.com/mytourbook/mytourbook/assets/50132146/1d0b378c-7bd8-4c9c-b46c-b781fed65f2f)


There is also no lap in the file, but it seems not mandatory.

Here are fit SDK explanations : 
![image](https://github.com/mytourbook/mytourbook/assets/50132146/c2bde55f-a555-4d58-b541-3d224f11afe1)

![image](https://github.com/mytourbook/mytourbook/assets/50132146/b39d9015-b15a-4072-bad7-b08acc0ac9a7)

![image](https://github.com/mytourbook/mytourbook/assets/50132146/7a65ae8d-18f9-417b-8d18-02b8b4506a8a)

**Comment by FJBDev** [comment]
> Hi, To conclude my analysis about fit files export from MTB, i discover a new "tool" in the fit SDK : FitTestTool.jar I submitted MTB fit file to it, and it founds 4 main "problems" . Those problems seems currently not making any issues with Suunto Import for my case, but it may be worth having a look at them, to decide if it worth MTB to try to align. Here are those "problems" 

@mff73 Thanks for spending time on that. I will be address those issues when I implement the fix for the Suunto 3D replay.
Does that mean that after those changes, the FIT Export from MT will be "perfect" 😄  ???
> 
>     1. FileId message : no manufacturer id

For that one, given that MyTourbook is not a device such as a watch, a gps unit... or a brand like Suunto/Garmin.... I would argue that we just can't add a manufacturer id. For example, I'm pretty sure that a FIT file generated by other apps like GoldenCheetah don't have a Manufacturer id.... but maybe I am wrong ?
> 
>     2. Activity message : no timestamp, no local timestamp

Will add.

>     3. Activity message : no total_timer_time

Will add.

> 
>     4. Session message : no timestamp

Will add.

**Comment by mff73** [comment]
> @mff73 Thanks for spending time on that. I will be address those issues when I implement the fix for the Suunto 3D replay. Does that mean that after those changes, the FIT Export from MT will be "perfect" 😄 ???

I don't know 😁
What i may begin to understand now also, there seems to be stricts rules to be followed, and some others probably more : facultative or dependant on each plateform which is using/decoding those fit files.
Currently, those last """""issues""""" seems not to be blocking anything, i just found this fittesttool and thought it helped debugging 3Dreplay in Suunto (but it was not the case, root cause was elsewhere 😊)

Btw, thanks for all.

**Comment by FJBDev** [comment]
After spending some time implementing your finding, I am pleased to say that the fit export is now correct and working  when replaying in Suunto App !

Attached is your original file [2023-09-23T12-51-export MTB.zip](https://github.com/mytourbook/mytourbook/files/12839965/2023-09-23T12-51-export.MTB.zip) exported in FIT file with my fix in MT

I still need to do the other changes you requested but I think it's a good base to start with
Let me know

[AfterFix.zip](https://github.com/mytourbook/mytourbook/files/14809129/AfterFix.zip)

**Comment by mff73** [comment]
Super, thanks.
I am confirming that the Suunto replay is working : timer_trigger event is well at the correct place 👍
Did you remove all data recorded during the pause (like HR/temperature)? It was one of the thinking we had during investigations, but it was not the root cause, and it is not necessary : we are just loosing data in exported fit files.
Are you planning to include them "again" (data during pauses) ?

**Comment by FJBDev** [comment]
> Did you remove all data recorded during the pause (like HR/temperature)? It was one of the thinking we had during investigations, but it was not the root cause, and it is not necessary : we are just loosing data in exported fit files. Are you planning to include them "again" (data during pauses) ?

I did because if i didnt remove them then it was not working.

**Comment by mff73** [comment]
strange, because now i am knowing the root cause, here is my original fit file, with just moving the timer event stop and start at their right places, and it is working well with Suunto replay, and data during the pause are still there.
here is the fit file and the csv from which i made the change with fitcsvtool.
I imported this fit file to suunto using runalyze.
I also re-add the missing data in your fit file after-fix, and it is working as well.

So, it seems that data during pause are not the issue if it was not working :(

[2023-09-23T12-51-export MTB just move timer event.zip](https://github.com/mytourbook/mytourbook/files/14812810/2023-09-23T12-51-export.MTB.just.move.timer.event.zip)

**Reply by FJBDev** [reply]
After more investigation, I think it's because I had misplaced the start event after the initial pause stop event before the data that should be located during the pause. I will try again and let you know...
But let me ask you this ? Is it useful to export the data during the pauses since, well, by definition, nothing happens when the device is paused ?

Also, I need some clarification for the below items
>     2. Activity message : no timestamp, no local timestamp

What type of timestamp ? the time the file was created ? the time that the activity began ? ended ?

> 
>     4. Session message : no timestamp

Same question as above

Thanks for your patience !

**Comment by mff73** [comment]
> After more investigation, I think it's because I had misplaced the start event after the initial pause stop event before the data that should be located during the pause. I will try again and let you know... 

> Thanks for your patience !

My patience is not the point to be thanked here, Your patience to work and spent Your time to improve MTB at such a level of feature must be thanked  : THANKS 😊

>But let me ask you this ? Is it useful to export the data during the pauses since, well, by definition, nothing happens when the device is paused ?

Let me answer by another question :) Why MTB is importing such "paused" data when they are present in any input files and why some devices are recording such data :), if it is to loose them at a potential re-export ;) ?
In terms of usages of such data, it may depend on each person use case : for example, in terms of training, one don't want to count pause (time/distance/etc...) related to active training phases, but could eventually see one's HR behaviour during the pause. Or temperature evolution during also a big picnic pause, or .........., one who is pausing during a whole night and also need/want to see some data (which are recorded --> so come back to my question : if data are there initially, MTB is saving them, displaying them, but doesn't re-export them, it would be quite not logic (to me)).

> Also, I need some clarification for the below items
> > 2. Activity message : no timestamp, no local timestamp
> What type of timestamp ? the time the file was created ? the time that the activity began ? ended ?
> > 4. Session message : no timestamp
> Same question as above

i don't precisely know, as i am just reading the fitsdk knowledge, i am learning with you 😊
Out of total logic, i opened few of my suunto fit files, and timestamps in activity and sesssion messages are the same as time_created.
What seems almost sure to me, is that local_timestamp in the activity message must be the same activity/timestamp source but in local area : if time_created choosen, then activity message will have timestamp=time_created AND local_timestamp=local_time_created --> to any potential decoders to be able to calculate the local time difference.
Isn't MTB using it any where ?
Btw, i just think these timestamps are there for any further potential decoding usages and calculations, and except timezone offset i don't see any usage of the timestamp value in session/activity messages.


And last thing discovered, it seems that MTB doesn't put file_id message in first place : 
![image](https://github.com/mytourbook/mytourbook/assets/50132146/bb4f4f1b-769a-45fc-992e-e2588275d124)

![image](https://github.com/mytourbook/mytourbook/assets/50132146/eebd3e39-d803-4dd8-8164-83a1a753ad3f)

**Comment by FJBDev** [comment]
> Let me answer by another question :) Why MTB is importing such "paused" data when they are present in any input files and why some devices are recording such data :), if it is to loose them at a potential re-export ;) ? In terms of usages of such data, it may depend on each person use case : for example, in terms of training, one don't want to count pause (time/distance/etc...) related to active training phases, but could eventually see one's HR behaviour during the pause. Or temperature evolution during also a big picnic pause, or .........., one who is pausing during a whole night and also need/want to see some data (which are recorded --> so come back to my question : if data are there initially, MTB is saving them, displaying them, but doesn't re-export them, it would be quite not logic (to me)).

I can add the data recorded during pauses but I am confused that now you want it because [above ](https://github.com/mytourbook/mytourbook/discussions/1194#discussioncomment-7248372) you were saying that this data was "fake" so i assume you didn't want "fake" data for a given FIT file ?

> Maybe also link to the fact that MTB seems to create "fake" GPS points during pause (like a straight line between beginning of pause and end of pause).


> i don't precisely know, as i am just reading the fitsdk knowledge, i am learning with you 😊 Out of total logic, i opened few of my suunto fit files, and timestamps in activity and sesssion messages are the same as time_created. What seems almost sure to me, is that local_timestamp in the activity message must be the same activity/timestamp source but in local area : if time_created choosen, then activity message will have timestamp=time_created AND local_timestamp=local_time_created --> to any potential decoders to be able to calculate the local time difference. Isn't MTB using it any where ? Btw, i just think these timestamps are there for any further potential decoding usages and calculations, and except timezone offset i don't see any usage of the timestamp value in session/activity messages.

Ok, I can add those timestamps. Hopefully they respect what the FIT SDK expects

> 
> And last thing discovered, it seems that MTB doesn't put file_id message in first place :

Here is the [official](All FIT files must contain a single File Id message. The File Id message identifies the intent of the FIT file through the Type field. The File Id message should be the first message in the file. Device serial number, managed by manufacturer.)definition I found for the file Id:

>All FIT files must contain a single File Id message. The File Id message identifies the intent of the FIT file through the Type field. The File Id message should be the first message in the file. Device serial number, managed by manufacturer.

Like I mentioned [above](https://github.com/mytourbook/mytourbook/discussions/1194#discussioncomment-8571600), I don't think it applies to an app that is not a "device" but maybe I am wrong. And if I am, what is the official Id of Mytourbook that Garmin assigned to us ? do you know ?

>For that one, given that MyTourbook is not a device such as a watch, a gps unit... or a brand like Suunto/Garmin.... I would argue that we just can't add a manufacturer id. For example, I'm pretty sure that a FIT file generated by other apps like GoldenCheetah don't have a Manufacturer id.... but maybe I am wrong ?

**Comment by FJBDev** [comment]
Attached is your exported original file generated with my latest code.
I believe  I addressed the issue you originally found (Suunto App Replay) + the requirements you mentioned when validating the FIT file with the Validator provided in the SDK.
Note on that utility: I "played" with it too and sorry to break the news to you but the Suunto files are faaaarrrrr from "perfect". See below all the failures with a file generated by my Suunto Vertical

[afterFix.zip](https://github.com/mytourbook/mytourbook/files/14844803/afterFix.zip)

```
Activity Message total_timer_time is Equal to the Sum of Session Messages total_timer_time Values - Level: REQUIRED Status: FAILED
	Activity Message contains a null total_timer_time
Session Message Valid Message Index - Level: REQUIRED Status: FAILED
	One or more Session messages contain a null Message Index value.
Lap Message Exists - Level: REQUIRED Status: FAILED
Device Info Message Device Index is Valid - Level: REQUIRED Status: FAILED
	One or more Device Info messages contain a null Device Index value
```

**Comment by mff73** [comment]
Thanks for all. MTB is now the best fit file creator. I am really happy.

About data during pause, when i wrote "fake" one day, it was to say "fake GPS points" = "interpolated GPS lat/lon" during the pause. I understood that MTB need lat/lon for its time slices to support other kind of data (HR, temp, etc....) even during pause.

**Reply by FJBDev** [reply]
@mff73 
You're welcome.
Do you want a test version of MT so you can "put it to the wringer" and make sure it's ready for the next release ?

**Reply by mff73** [reply]
It would be pleasure to test it for sure.

**Reply by FJBDev** [reply]
@mff73 

You can download the test version of MT [here](https://www.dropbox.com/scl/fi/q689k6er6stgq1aylytwy/mytourbook-24.TEST.0-win-setup.exe?rlkey=bdodu3fevqrm40jigdwmr0mv0&dl=0).

Thanks for testing. Once you validate my changes, I'll put them in a PR for the next MT version.

:warning: **VERY IMPORTANT** :warning:: You MUST NOT use your current MyTourbook database because any changes in this test (development) version, e.g. tour database version, can cause lots of troubles when you install again the current official release version.
If you REALLY want to use your current MyTourbook database, PLEASE BACKUP your database first as this build is **NOT** an official release. That way, once you uninstall it and install the official release again, you can restore your data unaltered.

See the [official documentation](http://mytourbook.sourceforge.net/mytourbook/index.php/documentation/backup) here on how to backup the database

Let me know and thanks for your testing

**Reply by mff73** [reply]
Cool, thanks.
Test version installed with separated db 😊
It is working well with one or two examples (recent ones).
I will "presser le citron" to stress this feature and see if some use cases are not working (longer pauses, or many many pauses, (like auto pause (which i don't do that much, but i may have some examples)).
Do you have any timeframe, if you want/need to release "soon" or not ?

**Reply by FJBDev** [reply]
>Do you have any timeframe, if you want/need to release "soon" or not ?

I don't have any timeframe at the moment as MT doesn't follow any schedule

**Reply by mff73** [reply]
Hello,
Each time i am using this new feature in the test version...... IT IS WORKING WELL 🥳
It is definitively really nice.
Thanks !

**Comment by FJBDev** [comment, marked as answer]
> Hello, Each time i am using this new feature in the test version...... IT IS WORKING WELL 🥳 It is definitively really nice. Thanks !

Thanks for testing (and also finding the original issue!), i will put the fix in the next version of MT.
